### PR TITLE
Allow batch editing of molecules: removal only

### PR DESCRIPTION
**Pull request by greglandrum**
The idea here is to make it simpler and less error prone to remove atoms/bonds from a molecule.

In the current code if you want to remove more than one atom you need to be sure to do the removals ordered by decreasing atom index. Code needs to use a pattern like this:
```
aidsToRemove=[]
for atom in rwmol.GetAtoms():
  if we_want_to_remove(atom):
    aidsToRemove.append(atom.GetIdx())

for aid in sorted(aidsToRemove,reverse=True):
  rwmol.RemoveAtom(aid)
```
with the new API this can be simplified to:
```
rwmol.BeginBatchEdit()
for atom in rwmol.GetAtoms():
  if we_want_to_remove(atom):
    rwmol.RemoveAtom(atom)
rwmol.CommitBatchEdit()
```
or, using a context manager:
```
with Chem.RWMol(mol) as rwmol:
  for atom in rwmol.GetAtoms():
    if we_want_to_remove(atom):
      rwmol.RemoveAtom(atom)
```

Both are simpler and, I think, less error prone.

Later we can also do some optimization work in here and make it more efficient to make these kinds of edits, but the idea with this PR is to establish the API.

### CODE REVIEW

**Comment by d-b-w** [comment]
I feel like

In C++, an RAII class seems like the right way to do this:
    
    {
       EditTransaction scoped(rwmol);
       for (auto atom: rwmol.GetAtoms()) {
          if (we_want_to_remove(atom)) {
            rwmol.RemoveAtom(atom)
          }
       }
     }

In Python, there is an even more natural syntax: a context manager:

    with rwmol.EditTransaction()
        for atom in rwmol.GetAtoms():
            if we_want_to_remove(atom):
                rwmol.RemoveAtom(atom)

**Comment by d-b-w** [comment]
(this is a very handy idea, by the way)

**Comment by greglandrum** [comment]
@d-b-w : I like both of those ideas. I'd want to keep the granular `beginBatchEdit()/commitBatchEdit()` methods available too, but I'll look into implementing the context manager and the scoped class in C++

**Comment by bp-kelley** [comment]
@greglandrum I think having the ability to rollback edits might add ton of complexity, you have to deal with conformers etc.

The simple solution is to copy the molecule try the edits and revert, but this might be a big performance hit.

**Comment by greglandrum** [comment]
@bp-kelley With the current implementation there’s no penalty whatsoever: it doesn’t make any change at all to the molecule until you commit the edits.
As promised, this is different from what you are probably expecting based on experience with other toolkits.
I’m not trying to reproduce what other toolkits do, just make the RDKit simpler (and a bit safer) to use and lay the groundwork for some performance improvements in the future.

**Comment by bp-kelley** [comment]
@greglandrum Yeah, I was projecting to the daylight method where you only know the molecule is borked (i.e. not sanitizable) when you close the edits.

This actually makes me want a predicate to remove atoms that doesn't change the atom index during the search.

```
RWMol::removeAtoms( [](const Atom &atom) {
   return atoms_to_remove.find(atom.getIdx()) != atoms_to_remove.end();
  }
)
```

File: Code/GraphMol/Abbreviations/Abbreviations.cpp [hunk 1]
```diff
@@ -76,11 +76,13 @@ void applyMatches(RWMol& mol, const std::vector<AbbreviationMatch>& matches) {
       mol.addBond(oaidx, connectIdx, Bond::BondType::SINGLE);
     }
   }
-  for (unsigned int i = toRemove.size(); i > 0; --i) {
-    if (toRemove[i - 1]) {
-      mol.removeAtom(i - 1);
+  mol.beginBatchEdit();
```
Comment (bp-kelley) [inline Code/GraphMol/Abbreviations/Abbreviations.cpp:79]: As you know I really like RAII, perhaps we could use a C++ context manager as well

```
BatchEdit editor(mol);
for(auto &atm : mol.atoms()) {
    mol.removeAtom(atm)
}
```

```
BatchEdit editor(mol);
try {
for(auto &atm : mol.atoms()) {
      mol.removeAtom(atm)
} catch (...) {
  editor.cancel();
}
}
```

I think this is a lot less error prone than missing a mol.commitBatchEdit();

File: Code/GraphMol/ROMol.cpp [hunk 2]
```diff
@@ -102,6 +102,17 @@ void ROMol::initFromOther(const ROMol &other, bool quickCopy, int confId) {
     d_stereo_groups.emplace_back(otherGroup.getGroupType(), std::move(atoms));
   }
 
+  if (other.dp_delAtoms) {
+    dp_delAtoms.reset(new boost::dynamic_bitset<>(*other.dp_delAtoms));
```
Comment (bp-kelley) [inline Code/GraphMol/ROMol.cpp:106]: I'm not sure I would copy over the editing state.  This seems really dangerous.

File: Code/GraphMol/RWMol.cpp [hunk 3]
```diff
@@ -448,4 +466,36 @@ unsigned int RWMol::finishPartialBond(unsigned int atomIdx2, int bondBookmark,
   return addBond(bsp->getBeginAtomIdx(), atomIdx2, bondType);
 }
 
+void RWMol::beginBatchEdit() {
+  if (dp_delAtoms || dp_delBonds) {
+    BOOST_LOG(rdWarningLog) << "batchEdit mode already enabled, ignoring "
```
Comment (bp-kelley) [inline Code/GraphMol/RWMol.cpp:471]: I think this should be an exception.  Something is wrong in the code and you should be forced to with this as an error.

File: Code/GraphMol/Wrap/rough_test.py [hunk 4]
```diff
@@ -6382,6 +6385,52 @@ def testContextManagers(self):
 
     RDLogger.EnableLog('rdApp.*')
 
+  def testBatchEdits(self):
+    mol = Chem.MolFromSmiles("C1CCCO1")
+
+    for rwmol in [Chem.EditableMol(mol), Chem.RWMol(mol)]:
+      rwmol.BeginBatchEdit()
+      rwmol.RemoveAtom(2)
```
Comment (bp-kelley) [inline Code/GraphMol/Wrap/rough_test.py:6393]: Need to check copying a mol in the middle of removing stuff.

File: Code/GraphMol/catch_graphmol.cpp [hunk 5]
```diff
@@ -1483,6 +1483,74 @@ TEST_CASE("needsHs function", "[chemistry]") {
   }
 }
 
+TEST_CASE("batch edits", "[editing]") {
+  SECTION("removeAtom") {
+    auto m = "C1CCCO1"_smiles;
+    REQUIRE(m);
+    m->beginBatchEdit();
+    m->removeAtom(2);
```
Comment (bp-kelley) [inline Code/GraphMol/catch_graphmol.cpp:1491]: Need to check copying a molecule in the middle of removing stuff.

**Review by bp-kelley — CHANGES_REQUESTED** [review]

File: Code/GraphMol/ROMol.cpp (hunk 2 above)
Comment (greglandrum) [reply Code/GraphMol/ROMol.cpp:106]: I don't see how there's any danger here given that no changes have actually been made to the molecule.

**Pull request merged**